chalk-integration/src/lowering/program_lowerer.rs

```diff
@@ -0,0 +1,428 @@
+use chalk_ir::cast::Cast;
+use chalk_ir::{
+    self, AdtId, AssocTypeId, BoundVar, ClosureId, DebruijnIndex, FnDefId, ForeignDefId, ImplId,
+    OpaqueTyId, TraitId, TyKind, VariableKinds,
+};
+use chalk_parse::ast::*;
+use chalk_solve::rust_ir::{
+    self, Anonymize, AssociatedTyValueId, OpaqueTyDatum, OpaqueTyDatumBound,
+};
+use rust_ir::IntoWhereClauses;
+use std::collections::{BTreeMap, HashSet};
+use std::sync::Arc;
+use string_cache::DefaultAtom as Atom;
+
+use super::{env::*, Lower, LowerParameterMap, LowerWithEnv, FIXME_SELF};
+use crate::error::RustIrError;
+use crate::program::Program as LoweredProgram;
+use crate::RawId;
+use crate::{interner::ChalkIr, TypeKind, TypeSort};
+
+#[derive(Default)]
+pub(super) struct ProgramLowerer {
+    next_item_index: u32,
+
+    associated_ty_lookups: AssociatedTyLookups,
+    associated_ty_value_ids: AssociatedTyValueIds,
+    adt_ids: AdtIds,
+    fn_def_ids: FnDefIds,
+    closure_ids: ClosureIds,
+    trait_ids: TraitIds,
+    auto_traits: AutoTraits,
+    opaque_ty_ids: OpaqueTyIds,
+    adt_kinds: AdtKinds,
+    fn_def_kinds: FnDefKinds,
+    closure_kinds: ClosureKinds,
+    trait_kinds: TraitKinds,
+    opaque_ty_kinds: OpaqueTyKinds,
+    object_safe_traits: HashSet<TraitId<ChalkIr>>,
+    foreign_ty_ids: ForeignIds,
+}
+
+impl ProgramLowerer {
+    pub fn next_item_id(&mut self) -> RawId {
+        let index = self.next_item_index;
+        self.next_item_index += 1;
+        RawId { index }
+    }
+
+    /// Create ids for associated type declarations and values
+    pub fn find_associated_types(
```

Comment:
Naming could be better here. Maybe `extract_associated_types` or `insert_associated_types`.